src/Adapter/MSTest.TestAdapter/Discovery/TypeEnumerator.cs

```diff
@@ -147,14 +149,40 @@ internal UnitTestElement GetTestFromMethod(MethodInfo method, bool isDeclaredInT
                     method.DeclaringType.Assembly);
         }
 
+        // PERF: AssemblyEnumerator would later try to figure if this is a data driven test method
+        // by creating a TestMethodInfo. This is really expensive for discovery, and should be done only
+        // when we are a data driven test.
+        // Instead we check if we have any data driven attribute here, and skip creating TestMethod info
+        // in AssemblyEnumerator.
+        //
+        // The foreach below could move to ReflectHelper, but this is only place where we want to check if
+        // an attribute implements something different than an Attribute derived class. So instead we just
+        // grab all attributes, and do the check ourselves.
+        DynamicDataType dynamicDataType = DynamicDataType.None;
+        foreach (Attribute attribute in _reflectHelper.GetDerivedAttributes<Attribute>(method, inherit: true))
+        {
+            if (AttributeComparer.IsDerived<ITestDataSource>(attribute))
+            {
+                dynamicDataType = DynamicDataType.ITestDataSource;
+                break;
+            }
+            else if (AttributeComparer.IsDerived<DataSourceAttribute>(attribute))
+            {
+                dynamicDataType = DynamicDataType.DataSourceAttribute;
+                break;
+            }
+        }
+
+        testMethod.DataType = dynamicDataType;
```

Comment:
Create a method for this so we are left with only `testMethod.DataType = XXX`

Reply:
done.